backend/pkg/modules/internal/differ/spec_diffs.go

```diff
@@ -0,0 +1,277 @@
+)
+
+const (
+	ModuleName = "differ"
+)
+
+type pluginDiffer struct {
+	httpHandler http.Handler
+
+	apiIDToDiffs map[uint]map[[32]byte]global.Diff
+	totalEvents  int
+
+	accessor core.BackendAccessor
+	sync.RWMutex
+}
+
+//nolint:gochecknoinits // was needed for the module implementation of ApiClarity
+func init() {
+	core.RegisterModule(newDiffer)
+}
+
+//nolint:ireturn,nolintlint // was needed for the module implementation of ApiClarity
+func newDiffer(ctx context.Context, accessor core.BackendAccessor) (core.Module, error) {
+	// Use default values
+	plugin := pluginDiffer{
+		httpHandler:  nil,
+		accessor:     accessor,
+		apiIDToDiffs: make(map[uint]map[[32]byte]global.Diff),
+		totalEvents:  0,
+	}
+
+	// TODO do we want to recover some state here?
+
+	go plugin.StartDiffsSender(ctx)
+
+	return &plugin, nil
+}
+
+func (p *pluginDiffer) Name() string {
+	return ModuleName
+}
+
+func (p *pluginDiffer) EventNotify(ctx context.Context, event *core.Event) {
+	var reconstructedDiff *_spec.APIDiff
+	var providedDiff *_spec.APIDiff
+	var err error
+
+	log.Infof("Got new event notification. event=%+v", event)
+
+	apiEvent := event.APIEvent
+	specKey := _speculator.GetSpecKey(apiEvent.HostSpecName, strconv.Itoa(int(apiEvent.DestinationPort)))
+	speculatorAccessor := p.accessor.GetSpeculatorAccessor()
+
+	if !speculatorAccessor.HasProvidedSpec(specKey) && !speculatorAccessor.HasApprovedSpec(specKey) {
+		log.Debugf("No diffs to calculate")
+		return
+	}
+
+	speculatorTelemetry := speculatorutils.ConvertModelsToSpeculatorTelemetry(event.Telemetry)
+
+	reconstructedDiffType := models.DiffTypeNODIFF
+	providedDiffType := models.DiffTypeNODIFF
+	if speculatorAccessor.HasProvidedSpec(specKey) {
+		// calculate diffs base on the event
+		providedDiff, err = speculatorAccessor.DiffTelemetry(speculatorTelemetry, _spec.DiffSourceProvided)
+		if err != nil {
+			log.Errorf("Failed to diff telemetry against provided spec: %v", err)
+			return
+		}
+		if err := setAPIEventProvidedDiff(apiEvent, providedDiff); err != nil {
+			log.Errorf("Failed to set api event provided diff: %v", err)
+			return
+		}
+		providedDiffType = convertAPIDiffType(providedDiff.Type)
+	} else if speculatorAccessor.HasApprovedSpec(specKey) {
+		// calculate diffs base on the event
+		reconstructedDiff, err = speculatorAccessor.DiffTelemetry(speculatorTelemetry, _spec.DiffSourceReconstructed)
+		if err != nil {
+			log.Errorf("Failed to diff telemetry against approved spec: %v", err)
+			return
+		}
+		if err := setAPIEventReconstructedDiff(apiEvent, reconstructedDiff); err != nil {
+			log.Errorf("Failed to set api event reconstructed diff: %v", err)
+			return
+		}
+		reconstructedDiffType = convertAPIDiffType(reconstructedDiff.Type)
+	}
+
+	apiEvent.SpecDiffType = getHighestPrioritySpecDiffType(providedDiffType, reconstructedDiffType)
+
+	// save api event with diffs in db
+	if err := p.accessor.UpdateAPIEvent(ctx, apiEvent); err != nil {
```

Comment:
Modules should not modify events, but only add annotations to event. Is it possible for the diff module to store the diffs as annotations?